crypto/rand/drbg_lib.c

```diff
@@ -84,169 +59,178 @@ int RAND_DRBG_set(DRBG_CTX *dctx, int nid, unsigned int flags)
 
 /*
  * Allocate memory and initialize a new DRBG.  The |parent|, if not
- * NULL, will be used to auto-seed this DRBG_CTX as needed.
+ * NULL, will be used to auto-seed this RAND_DRBG as needed.
  */
-DRBG_CTX *RAND_DRBG_new(int type, unsigned int flags, DRBG_CTX *parent)
+RAND_DRBG *RAND_DRBG_new(int type, unsigned int flags, RAND_DRBG *parent)
 {
-    DRBG_CTX *dctx = OPENSSL_zalloc(sizeof(*dctx));
+    RAND_DRBG *drbg = OPENSSL_zalloc(sizeof(*drbg));
 
-    if (dctx == NULL) {
+    if (drbg == NULL) {
         RANDerr(RAND_F_RAND_DRBG_NEW, ERR_R_MALLOC_FAILURE);
         return NULL;
     }
 
-    dctx->parent = parent;
-    if (RAND_DRBG_set(dctx, type, flags) < 0) {
-        OPENSSL_free(dctx);
-        return NULL;
+    drbg->parent = parent;
+    if (RAND_DRBG_set(drbg, type, flags) < 0)
+        goto err;
+
+    if (parent != NULL) {
+        if (!RAND_DRBG_set_callbacks(drbg, drbg_entropy_from_parent,
+                                     drbg_release_entropy,
+                                     NULL, NULL)
+                /*
+                 * Add in our address.  Note we are adding the pointer
+                 * itself, not its contents!
+                 */
+                || !RAND_DRBG_instantiate(drbg,
+                                          (unsigned char*)&drbg, sizeof(drbg)))
+            goto err;
     }
-    return dctx;
+
+    return drbg;
+
+err:
+    OPENSSL_free(drbg);
+    return NULL;
 }
 
 /*
- * Uninstantiate |dctx| and free all memory.
+ * Uninstantiate |drbg| and free all memory.
  */
-void RAND_DRBG_free(DRBG_CTX *dctx)
+void RAND_DRBG_free(RAND_DRBG *drbg)
 {
-    if (dctx == NULL)
+    if (drbg == NULL)
         return;
 
-    ctr_uninstantiate(dctx);
-    CRYPTO_free_ex_data(CRYPTO_EX_INDEX_DRBG, dctx, &dctx->ex_data);
+    ctr_uninstantiate(drbg);
+    CRYPTO_free_ex_data(CRYPTO_EX_INDEX_DRBG, drbg, &drbg->ex_data);
 
     /* Don't free up default DRBG */
-    if (dctx == RAND_DRBG_get_default()) {
-        memset(dctx, 0, sizeof(DRBG_CTX));
-        dctx->nid = 0;
-        dctx->status = DRBG_STATUS_UNINITIALISED;
+    if (drbg == &rand_drbg) {
+        OPENSSL_cleanse(drbg, sizeof(*drbg));
+        drbg->nid = 0;
```

Comment:
well it's an impossible condition since app code can never get to rand_drbg :)  But I made it be a no-op and pointed to rand_cleanup_int() in a comment.  Thanks.